File: typescript/src/days/day04.ts
type Ticket = {
    id: number,
    wins: number[],
    mine: number[],
    copies: number
}

export function day04(data: string) {
    const tickets = parseTickets(data);
    const result1 = part1(tickets);
    const result2 = part2(tickets);

    console.log('Total number of points is', result1);
    console.log('Total number of tickets is', result2);
}

function parseTickets(data: string): Ticket[] {
    return data
        .split('\n')
        .map(line => {
            const [idPart, winningPart, myParts] = line.split(/[:|]/g);

            const id = +idPart.match(/\d+/g)[0];
            const wins = winningPart.match(/\d+/g).map(Number);
            const mine = myParts.match(/\d+/g).map(Number);
            return { id, wins, mine, copies: 1 };
        })
}

function part1(tickets: Ticket[]) {
    const scores = tickets
        .map(({ wins, mine }) => {
            const winningNumbers = mine.filter(m => wins.includes(m)).length;
            return winningNumbers === 0 ? 0 : (2 ** (winningNumbers - 1));
        })

    return scores.reduce((a, b) => a + b);
}

function part2(tickets: Ticket[]) {
    let count = 0;

    for(let i = 0; i < tickets.length; i++) {
        const { wins, mine, copies } = tickets[i];
        count += copies;

        const winningNumbers = mine.filter(m => wins.includes(m)).length;
        for(let j = 1; j <= winningNumbers; j++) {
            tickets[j+i].copies += copies
        }
    }
    
    return count;
}

File: go/day04/day04.go
package day04

import (
	"fmt"
	"os"
	"regexp"
	"slices"
	"strconv"
	"strings"
)

type Ticket struct {
	wins   []int
	mine   []int
	amount int
}

func Day04() {
	data, _ := os.ReadFile("../data/day04.txt")
	// data, _ := os.ReadFile("./test.txt")
	tickets := parseTickets(string(data))

	points := part1(tickets)
	totalAmount := part2(tickets)

	fmt.Println("\n===== Day 04 =====")
	fmt.Println("The total points are", points)
	fmt.Println("The total number of tickets is", totalAmount)
}

func part1(tickets []Ticket) int {
	points := 0
	for _, ticket := range tickets {
		score := 0
		for _, num := range ticket.mine {
			if slices.Contains(ticket.wins, num) {
				if score == 0 {
					score = 1
				} else {
					score *= 2
				}
			}
		}
		points += score
	}
	return points
}

func part2(tickets []Ticket) int {
	count := 0
	for i, ticket := range tickets {
		count += ticket.amount
		hits := 0
		for _, num := range ticket.mine {
			if slices.Contains(ticket.wins, num) {
				hits++
			}
		}
		for j := 1; j <= hits && j < len(tickets); j++ {
			tickets[i+j].amount += ticket.amount
		}
	}
	return count
}

func parseTickets(data string) []Ticket {
	tickets := make([]Ticket, 0)
	lines := strings.Split(data, "\n")
	for _, line := range lines {
		wins := getNumbers(strings.Split(strings.Split(line, ":")[1], "|")[0])
		mine := getNumbers(strings.Split(strings.Split(line, ":")[1], "|")[1])
		tickets = append(tickets, Ticket{wins, mine, 1})
	}
	return tickets
}

func getNumbers(s string) []int {
	numsRegex := regexp.MustCompile(`\d+`)
	numStrings := numsRegex.FindAllString(s, -1)
	nums := make([]int, 0)
	for _, numString := range numStrings {
		num, _ := strconv.Atoi(numString)
		nums = append(nums, num)
	}
	return nums
}
